checked_message_test.go

```diff
@@ -60,3 +60,140 @@ func TestCheckedMessageIsSingleUse(t *testing.T) {
+
+		`{"level":"info","msg":"elderberry","name":"A","i":7}`,
+		`{"level":"info","msg":"elderberry","name":"B","i":7}`,
+
+		// trees
+		`{"level":"info","msg":"fig","name":"A","i":8}`,
+		`{"level":"info","msg":"fig","name":"B","i":8}`,
+		`{"level":"info","msg":"fig","name":"C","i":8}`,
+		`{"level":"info","msg":"guava","name":"B","i":9}`,
+		`{"level":"info","msg":"guava","name":"C","i":9}`,
+		`{"level":"info","msg":"guava","name":"A","i":9}`,
+	}
+	withJSONLogger(t, opts(InfoLevel), func(logger Logger, buf *testBuffer) {
+		loga := logger.With(String("name", "A"))
+		logb := logger.With(String("name", "B"))
+		logc := logger.With(String("name", "C"))
+
+		tests := []struct {
+			cms        []*CheckedMessage
+			expectedOK bool
+			desc       string
+		}{
+			// not-ok base cases
+			{
+				cms:        []*CheckedMessage{nil},
+				expectedOK: false,
+				desc:       "nil",
+			},
+			{
+				cms: []*CheckedMessage{
+					nil,
+					loga.Check(DebugLevel, "nope"),
+				},
+				expectedOK: false,
+				desc:       "nil, A decline",
+			},
+			{
+				cms: []*CheckedMessage{
+					nil,
+					loga.Check(DebugLevel, "nope"),
+					logb.Check(DebugLevel, "nope"),
+				},
+				expectedOK: false,
+				desc:       "nil, A and B decline",
+			},
+
+			// singleton ok cases
+			{
+				cms: []*CheckedMessage{
+					nil,
+					loga.Check(InfoLevel, "apple"),
+				},
+				expectedOK: true,
+				desc:       "nil, A OK",
+			},
+			{
+				cms: []*CheckedMessage{
+					loga.Check(InfoLevel, "banana"),
+					logb.Check(DebugLevel, "banana"),
+				},
+				expectedOK: true,
+				desc:       "A OK, B decline",
+			},
+
+			// compound ok cases
+			{
+				cms: []*CheckedMessage{
+					loga.Check(InfoLevel, "cherry"),
+					logb.Check(InfoLevel, "cherry"),
+				},
+				expectedOK: true,
+				desc:       "A and B OK",
+			},
+
+			{
+				cms: []*CheckedMessage{
+					loga.Check(InfoLevel, "durian"),
+					nil,
+					logb.Check(InfoLevel, "durian"),
+				},
+				expectedOK: true,
+				desc:       "A OK, nil, B OK",
+			},
+
+			{
+				cms: []*CheckedMessage{
+					loga.Check(InfoLevel, "elderberry"),
+					logb.Check(InfoLevel, "elderberry"),
+					nil,
+				},
+				expectedOK: true,
+				desc:       "A OK, B OK, nil",
+			},
+
+			// trees
+			{
+				cms: []*CheckedMessage{
+					loga.Check(InfoLevel, "fig"),
+					logb.Check(InfoLevel, "fig").Chain(logc.Check(InfoLevel, "fig")),
+				},
+				expectedOK: true,
+				desc:       "A OK, ( B OK, C OK )",
+			},
+			{
+				cms: []*CheckedMessage{
+					logb.Check(InfoLevel, "guava").Chain(logc.Check(InfoLevel, "guava")),
+					loga.Check(InfoLevel, "guava"),
+				},
+				expectedOK: true,
+				desc:       "( B OK, C OK ), A OK",
+			},
+		}
+
+		for i, tt := range tests {
+			cm := tt.cms[0].Chain(tt.cms[1:]...)
+			assert.Equal(t, cm.OK(), tt.expectedOK, "%s: expected cm.OK()", tt.desc)
+			cm.Write(Int("i", i))
+		}
+		assert.Equal(t, expected, buf.Lines(), "expected output from MultiCheckedMessage")
```

Comment:
Could we include the expected output in the test structs, `Reset`ing the output buffer on each iteration? Might make it easier to identify the failing test case when we break something.